lib/cockpit/cockpit-auth-miq

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env ruby
+require "net/http"
+require "json"
+require "uri"
+require 'socket'
+
+MAX_AUTH_SIZE = 64 * 1024
+AUTH_FD = 10
+
+host = ARGV[-1]
+success=false
+url = URI.parse('http://localhost:3000/api/')
+if ENV['MQ_API_PATH']
+    url = URI.parse(ENV['MQ_API_PATH'])
+end
+
+ios = IO.for_fd(AUTH_FD)
+token = ios.readpartial(MAX_AUTH_SIZE)
+
+req = Net::HTTP::Get.new(url)
+req['X-Auth-Token'] = token
+res = Net::HTTP.start(url.hostname, url.port) {|http|
+  http.request(req)
+}
+
+json_resp = { 'error' => 'Unexpected Error' }
+
+if res.code == "200"
+    begin
+        success = true
+    rescue
+        json_resp['message'] = "Got invalid json response: " + res.body
+    end
+elsif res.code == "401"
+    json_resp['error'] = "authentication-failed"
+    json_resp['message'] = token
+    json_resp['auth-method-results'] = { "password" => "not-tried", "token" => "denied" }
+else
+    message = "Got invalid response: " + res.code
+end
+
+if !success
+    ios.syswrite(JSON.dump(json_resp))
+    exit(1)
+end
+
+# TODO: LOOKUP CREDENTIALS IN MIQ, SHOULD ONLY PROMPT IF NONE FOUND
+
+# OTHERWISE PROMPT FOR USER
+json_resp = { 'prompt' => 'Please provide a username:', 'echo' => 1, "message" => "Unable to determine the user to connect as." }
+ios.syswrite(JSON.dump(json_resp))
+user = ios.readpartial(MAX_AUTH_SIZE)
+
+# PROMPT FOR PW
+json_resp = { 'prompt' => 'Password:' }
+ios.syswrite(JSON.dump(json_resp))
+password = ios.readpartial(MAX_AUTH_SIZE)
+
+s1, s2 = UNIXSocket.socketpair(Socket::SOCK_SEQPACKET)
+s2.syswrite(password)
+
+trap("CHLD", "IGNORE")
```

Comment:
I suggest you stay away from trap since it doesn't play nicely with loggers or anything using mutexts.  Instead wrap your code in a begin/rescue SignalException block like we did below:

https://github.com/ManageIQ/manageiq/pull/2386